docs: add info about working with fixup commits (#39110)
angular/angular: docs/FIXUP_COMMITS.md

# Working with fixup commits

This document provides information and guidelines for working with fixup commits:
- [What are fixup commits](#about-fixup-commits)
- [Why use fixup commits](#why-fixup-commits)
- [Creating fixup commits](#create-fixup-commits)
- [Squashing fixup commits](#squash-fixup-commits)

[This blog post](https://thoughtbot.com/blog/autosquashing-git-commits) is also a good resource on the subject.

> ## <a name="about-fixup-commits"></a> What are fixup commits

Reviewer: Github creates auto-links for heading. Why not just use those?

Author: Custom anchors allows changing headings without breaking the links.
(Also, they are usually cleaner than headings - not in this case, because my headings are super-clear, but as a general rule 😛).

At their core, fixup commits are just regular commits with a special commit message:

> The first line of their commit message starts with `fixup! ` (notice the space after `!`) followed by the first line of the commit message of an earlier commit (it doesn't have to be the immediately preceeding one).

Reviewer: Add quotes to make the space stand out.

```suggestion
The first line of their commit message starts with "`fixup! `" (notice the space after `!`) followed by the first line of the commit message of an earlier commit (it doesn't have to be the immediately preceding one).
```

> The purpose of a fixup commit is to "fix up" an earlier commit.

Reviewer: This tautological sentence is a bit redundant. How about:

```suggestion
The purpose of a fixup commit is to modify an earlier commit.
```

I.e. it allows adding more changes in a new commit, but "marking" them as belonging to an earlier commit.
`Git` provides tools to make it easy to squash fixup commits into the original commit at a later time (see [below](#squash-fixup-commits) for details).

For example, let's assume you have added the following commits to your branch:

```
feat: first commit
fix: second commit
```

If you want to add more changes to the first commit, you can create a new commit with the commit message:
`fixup! feat: first commit`:

```
feat: first commit
fix: second commit
fixup! feat: first commit
```

## <a name="why-fixup-commits"></a> Why use fixup commits

So, when are fixup commits useful?

> Often times, when submitting a Pull Request, a reviewer might request changes.

Reviewer: ```suggestion
During the life of a Pull Request, a reviewer might request changes.
```

```sh
# Create a fixup commit to fix up the last commit on the branch:
git commit --fixup HEAD ...

# Create a fixup commit to fix up commit with SHA <COMMIT_SHA>:
git commit --fixup <COMMIT_SHA> ...
```

## <a name="squash-fixup-commits"></a> Squashing fixup commits

As mentioned above, the merge script will [automatically squash](../dev-infra/pr/merge/strategies/autosquash-merge.ts) fixup commits.
However, sometimes you might want to manually squash a fixup commit.

### Rebasing to squash fixup commits

The easiest way to re-order and squash any commit is via [rebasing interactively](https://git-scm.com/docs/git-rebase#_interactive_mode). You move a commit right after the one you want to squash it into in the rebase TODO list and change the corresponding action from `pick` to `fixup`.

`Git` can do all these automatically for you if you pass the `--autosquash` option to `git rebase`.
See the [`git` docs](https://git-scm.com/docs/git-rebase#Documentation/git-rebase.txt---autosquash) for more details.

> ### Configuring `git` to auto-squash by default

Reviewer: I would not recommend setting this default. When working with PRs I often rebase interactively but do not want to autosquash. For example, if a PR was built on top of another one, and you want to remove those commits when the upstream PR has been merged to master. Or when re-wording commit messages in a PR after a review finds typos.

Author: I am not recommeding it, but I wanted to have it mentioned.
<sub>(FWIW, I _do_ have this set to true 🤷‍♂️)</sub>

Author: Are you saying it would be better to remove this section or word it differently (how?) to make it not sound as if we are recommending it?

Reviewer: Definitely not remove it, but make it a bit less prominent so that it doesn't sound like a recommendation/requirement.

Author: Done.
